deepsubtree.go

```diff
@@ -0,0 +1,421 @@
+package iavl
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"strings"
+
+	ics23 "github.com/confio/ics23/go"
+)
+
+const (
+	// lengthByte is the length prefix prepended to each of the sha256 sub-hashes
+	lengthByte byte = 0x20
+)
+
+// Represents a IAVL Deep Subtree that can contain
+// a subset of nodes of an IAVL tree
+type DeepSubTree struct {
+	*MutableTree
+}
+
+// Traverses the nodes in the NodeDB that are part of Deep Subtree
+// and links them together using the populated left and right
+// hashes and sets the root to be the node with the given rootHash
+func (dst *DeepSubTree) BuildTree(rootHash []byte) error {
+	if dst.root == nil {
+		rootNode, rootErr := dst.ndb.GetNode(rootHash)
+		if rootErr != nil {
+			return fmt.Errorf("could not set root of deep subtree: %w", rootErr)
+		}
+		dst.root = rootNode
+	} else if !bytes.Equal(dst.root.hash, rootHash) {
+		return fmt.Errorf(
+			"deep Subtree rootHash: %s does not match expected rootHash: %s",
+			dst.root.hash,
+			rootHash,
+		)
+	}
+	nodes, traverseErr := dst.ndb.nodes()
+	if traverseErr != nil {
+		return fmt.Errorf("could not traverse nodedb: %w", traverseErr)
+	}
+	for _, node := range nodes {
+		pnode, err := dst.ndb.GetNode(node.hash)
+		if err != nil {
+			return err
+		}
+		err = dst.linkNode(pnode)
+		if err != nil {
+			return err
+		}
+	}
+	// Now that nodes are linked correctly, traverse again
+	// and set their keys correctly
+	for _, node := range nodes {
+		pnode, err := dst.ndb.GetNode(node.hash)
+		if err != nil {
+			return err
+		}
+		if pnode.leftNode != nil {
+			pnode.key = pnode.leftNode.getHighestKey()
+		}
+
+		if pnode.rightNode != nil {
+			pnode.key = pnode.rightNode.getLowestKey()
+		}
+	}
+
+	return nil
+}
+
+// Link the given node if it is not linked yet
+// If already linked, return an error in case connection was made incorrectly
+// Note: GetNode returns nil if the hash passed into it is empty which is expected
+// and err does not need to be checked
+func (dst *DeepSubTree) linkNode(node *Node) error {
+	if len(node.leftHash) > 0 {
+		if node.leftNode == nil {
+			node.leftNode, _ = dst.ndb.GetNode(node.leftHash)
+		} else if !bytes.Equal(node.leftNode.hash, node.leftHash) {
```

Comment:
Building off of my other comment, we should either explain why it is ok to return without an error and still a nil `node.leftNode` in this case, or we should update this to the following as an additional protection against mismatches.
```suggestion
		}
		if !bytes.Equal(node.leftNode.hash, node.leftHash) {
```

Reply:
This is an invalid code since if `node.leftNode == nil` is true, then `node.leftNode.hash` will crash.